src/bestcrypt_ve_fmt_plug.c

```diff
@@ -0,0 +1,456 @@
+/*
+ * JtR format to crack password protected BestCrypt v4 volume
+ * This implementation relies on findings from @trounce1 and @kholia 
+ * reverse-engineering work
+
+ * ----------------------------------------------------------------------------
+ * "THE BEER-WARE LICENSE" (Revision 42):
+ * <jean-christophe.delaunay@synacktiv.com> wrote this file. As long as you 
+ * retain this notice you can do whatever you want with this stuff. If we meet 
+ * some day, and you think this stuff is worth it, you can buy me a beer in 
+ * return
+ * ----------------------------------------------------------------------------
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_bestcrypt_ve;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_bestcrypt_ve);
+#else
+
+#include <string.h>
+#include <errno.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "arch.h"
+#include "misc.h"
+#include "common.h"
+#include "formats.h"
+#include "params.h"
+#include "options.h"
+#include "yescrypt/yescrypt.h"
+#include "aes.h"
+#include "twofish.h"
+#include "serpent.h"
+#include <openssl/camellia.h>
+#include "sha.h"
+#include "jumbo.h"
+
+#define FORMAT_NAME             "BestCrypt Volume Encryption v4 "
+#define FORMAT_LABEL            "Jetico BestCrypt"
+#define FORMAT_TAG              "$bcve$"
+#define FORMAT_TAG_LENGTH       (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "scrypt [algorithm]-256-CBC SHA-256"
+
+#define BENCHMARK_COMMENT       "scrypt parameters (32768, 16, 1)"
+#define BENCHMARK_LENGTH        0x507
+#define BINARY_SIZE             0
+#define PLAINTEXT_LENGTH        125
+#define SALT_SIZE               sizeof(*cur_salt)
+#define BINARY_ALIGN            sizeof(uint32_t)
```

Comment:
Switching from 0 to 1 gives the following error:

```
$ john --format=bestcryptve4 --test
Will run 8 OpenMP threads
Benchmarking: BestCryptVE4, BestCrypt Volume Encryption v4 (32768, 16, 1) [scrypt Salsa20/8 128/128 AVX, AES/TwoFish/Serpent/Camellia 256]... (8xOMP) FAILED (Using default binary() with a non-zero BINARY_SIZE)
```

I could not find some information regarding this constant, could you please explain me its purpose or point me out where should I look to understand things? (I saw a lot of formats also define it to 0).

Reply:
I meant `BINARY_ALIGN` should be 1. Not `BINARY_SIZE`, which was correctly 0.

You don't really need the below to complete this PR, but since you asked:

> point me out where should I look to understand things?

The formats interface and its related parameters is documented in `formats.h`. However, this would not fully explain things. Originally, the interface was designed for usual password hashes, where "binary" would have meant their binary representation (decoded from ASCII at loading). This format, like certain others introduced since 2011, is for what we call "non-hashes". With usual password hash formats, we compare computed hashes against loaded hashes. With non-hash formats, we might not have a target hash to compare against (and if so the binary would be zero-sized), but instead we store whatever other information we need for a password correctness check inside the salt structure (thus, we also call such formats "salt-only").

Reply:
Indeed, I misunderstood things regarding the concerned variable. Thanks for the explanation aside that!